include/ginkgo/core/distributed/neighborhood_communicator.hpp

```diff
@@ -0,0 +1,134 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_PUBLIC_CORE_DISTRIBUTED_NEIGHBORHOOD_COMMUNICATOR_HPP_
+#define GKO_PUBLIC_CORE_DISTRIBUTED_NEIGHBORHOOD_COMMUNICATOR_HPP_
+
+
+#include <ginkgo/config.hpp>
+
+
+#if GINKGO_BUILD_MPI
+
+#include <ginkgo/core/base/mpi.hpp>
+#include <ginkgo/core/distributed/collective_communicator.hpp>
+#include <ginkgo/core/distributed/index_map.hpp>
+
+
+namespace gko {
+namespace experimental {
+namespace mpi {
+
+
+/**
+ * A collective_communicator that uses a neighborhood topology.
+ *
+ * The neighborhood communicator is defined by a list of neighbors this
+ * rank sends data to and a list of neighbors this rank receives data from.
+ * No communication with any ranks that is not in one of those lists will
+ * take place.
+ */
+class neighborhood_communicator final : public collective_communicator {
```

Comment:
This is a polymorphic class, so according to our conventions, it should be in `CamelCase`. Same for its base class.